lib/at_cmd/at_cmd.c

```diff
@@ -58,6 +58,8 @@ static struct k_thread socket_thread;
 static at_cmd_handler_t notification_handler;
 static atomic_t shutdown_mode;
 
+/* Semaphore to guard the at_cmd init from simultaneous entry. */
+K_SEM_DEFINE(at_cmd_init_sem, 1, 1);
```

Comment:
I would suggest using mutex rather than semaphore in this case. Not an OS expert myself, so I may be on thin ice, but we are protecting something fundamentally single here. Mutex has the advantage of priority inheritance, and is also the mechanism used to protect other single resources in at_cmd.